language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/util/documentsymbol/SymbolUtil.java

```diff
@@ -0,0 +1,263 @@
+ *  under the License.
+ */
+package org.ballerinalang.langserver.util.documentsymbol;
+
+import io.ballerina.compiler.syntax.tree.ChildNodeList;
+import io.ballerina.compiler.syntax.tree.ClassDefinitionNode;
+import io.ballerina.compiler.syntax.tree.FunctionDefinitionNode;
+import io.ballerina.compiler.syntax.tree.IdentifierToken;
+import io.ballerina.compiler.syntax.tree.MetadataNode;
+import io.ballerina.compiler.syntax.tree.ModulePartNode;
+import io.ballerina.compiler.syntax.tree.Node;
+import io.ballerina.compiler.syntax.tree.ServiceDeclarationNode;
+import io.ballerina.compiler.syntax.tree.SimpleNameReferenceNode;
+import io.ballerina.compiler.syntax.tree.SyntaxKind;
+import io.ballerina.compiler.syntax.tree.SyntaxTree;
+import io.ballerina.projects.Module;
+import org.ballerinalang.langserver.commons.DocumentSymbolContext;
+import org.eclipse.lsp4j.DocumentSymbol;
+import org.eclipse.lsp4j.Position;
+import org.eclipse.lsp4j.Range;
+import org.eclipse.lsp4j.SymbolInformation;
+import org.eclipse.lsp4j.SymbolKind;
+import org.eclipse.lsp4j.SymbolTag;
+import org.eclipse.lsp4j.jsonrpc.messages.Either;
+
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * Finds Symbols within a document.
+ *
+ * @since 2.0.0
+ */
+public class SymbolUtil {
+
+    public static final Set<SyntaxKind> SUPPORTED_NODE_TYPES = Set.of(SyntaxKind.MODULE_PART,
+            SyntaxKind.FUNCTION_DEFINITION, SyntaxKind.METHOD_DECLARATION, SyntaxKind.SERVICE_DECLARATION,
+            SyntaxKind.CLASS_DEFINITION, SyntaxKind.RESOURCE_ACCESSOR_DEFINITION, SyntaxKind.OBJECT_METHOD_DEFINITION);
+
+    /**
+     * get document symbols given the context.
+     *
+     * @param context document symbol context.
+     * @return {@link List<Either<SymbolInformation, DocumentSymbol>>} generated document symbols.
+     */
+    public static List<Either<SymbolInformation, DocumentSymbol>> documentSymbols(DocumentSymbolContext context) {
+        List<Either<SymbolInformation, DocumentSymbol>> symbols = new ArrayList<>();
+        Optional<SyntaxTree> syntaxTree = context.currentSyntaxTree();
+        if (syntaxTree.isEmpty()) {
+            return symbols;
+        }
+        if (context.getHierarchicalDocumentSymbolSupport()) {
+            Optional<DocumentSymbol> rootSymbol = getDocumentSymbolForNode(syntaxTree.get().rootNode(),
+                    context, Collections.emptyList());
+            rootSymbol.ifPresent(documentSymbol -> symbols.add(Either.forRight(documentSymbol)));
+        } else {
+            List<DocumentSymbol> documentSymbolStore = new ArrayList<>();
+            getDocumentSymbolForNode(syntaxTree.get().rootNode(), context, documentSymbolStore);
+            documentSymbolStore.forEach(symbol -> symbols.add(Either.forRight(symbol)));
+        }
```

Comment:
Instead of the recursive call based implementation, shall we use node transformers? That will make the maintenance easier.